text_generation/causal_lm/cpp/speculative_decoding.cpp

```diff
@@ -0,0 +1,379 @@
+        ss_1.str("");
+        ss_2.str("");
+    }
+}
+
+void init_key_values(ov::InferRequest request, int kv_length) {
+    std::stringstream ss;
+    ss.clear();
+    
+    for (int i = 0; i < kv_length; i++) {
+        ss << "past_key_values." << i << ".key";
+        auto tensor_shape = request.get_tensor(ss.str()).get_shape();
+
+        request.set_tensor(ss.str(), ov::Tensor(ov::element::f32, {BATCH_SIZE, tensor_shape[1], 0, tensor_shape[3]}));
+        ss.str("");
+
+        ss << "past_key_values." << i << ".value";
+        request.set_tensor(ss.str(), ov::Tensor(ov::element::f32, {BATCH_SIZE, tensor_shape[1], 0, tensor_shape[3]}));
+        ss.str("");
+    }
+}
+
+ov::Tensor trimm_tensor(ov::Tensor tensor, uint64_t axis, uint64_t new_seq_len) {
+    // tensor shape = [batch, size_1, seq_len, size_2]
+    auto old_tensor_data = tensor.data<float>();
+    auto shape = tensor.get_shape();
+    auto size_1 = shape[1];
+    auto old_seq_len = shape[2];
+    auto size_2 = shape[3];
+    
+    // todo: assume batch is always 1, if different one more cycle is needed
+    // copy elements from the old to the new tensor
+    auto new_tensor = ov::Tensor{ov::element::f32, {BATCH_SIZE, size_1, new_seq_len, size_2}};
+    auto new_tensor_data = new_tensor.data<float>();
+    for (int i = 0; i < size_1; i++) {
+        for (int j = 0; j < new_seq_len; j++) {
+            for (int k = 0; k < size_2; k++) {
+                new_tensor_data[new_seq_len * size_2 * i +  size_2 * j + k] = old_tensor_data[old_seq_len * size_2 * i +  size_2 * j + k];
+            }
+        }
+    }
+    return new_tensor;
+}
+
+void update_kv_cache(ov::InferRequest request, uint64_t new_seq_len, int kv_length) {
+    // key/value shape = [batch, size_1, seq_len, size_2]
+    uint64_t seq_len_axis = 2;  // check if can be removed hardcoded value
+    
+    std::stringstream ss;
+    for (int i = 0; i < kv_length; i++) {
+        
+        for (auto name: {"key", "value"}) {
+            ss << "present." << i;
+            ss << (name == "key" ? ".key" : ".value");
+            auto tensor_name = ss.str();
+            
+            auto trimmed_tensor = trimm_tensor(request.get_tensor(tensor_name), seq_len_axis, new_seq_len);
+            request.set_tensor(tensor_name, trimmed_tensor);
+            ss.str("");
+        }
+    }
+}
+
+bool verbose = false;
+
+using namespace std;
+void print_accum_tokens(ov::InferRequest& detokenizer, std::vector<int64_t> tokens, string suffix) {
+    std::string text = detokenize(detokenizer, tokens);
+    stringstream ss;
+    for (const auto& token: tokens)
+        ss << token << " ";
+    cerr << suffix << ": ";
+    cerr << text << " | " << ss.str() << endl;
+}
+
+string tokens_to_string(ov::InferRequest& detokenizer, std::vector<int64_t> tokens) {
+    return detokenize(detokenizer, tokens);
+}
+
+int main(int argc, char* argv[]) try {
+    int draft_kv_size = 22;
+    int target_kv_size = 32;
```

Comment:
we need to extract if from model, take number of inputs, subtract 3 (input_ids, position_ids and mask) and divide by 2

Reply:
sure. This was a draft solution. I will generalize for ready for review version

Reply:
done. I use `inputs().size()`/`outputs().size()` instead of hardcoded values.